src/background.c

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright © 2019 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <glib/gi18n.h>
+#include <gio/gio.h>
+#include <gio/gdesktopappinfo.h>
+
+#include "background.h"
+#include "request.h"
+#include "permissions.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-utils.h"
+
+#define PERMISSION_TABLE "background"
+#define PERMISSION_ID "background"
+
+typedef struct _Background Background;
+typedef struct _BackgroundClass BackgroundClass;
+
+struct _Background
+{
+  XdpBackgroundSkeleton parent_instance;
+};
+
+struct _BackgroundClass
+{
+  XdpBackgroundSkeletonClass parent_class;
+};
+
+static XdpImplAccess *impl;
+static Background *background;
+
+GType background_get_type (void) G_GNUC_CONST;
+static void background_iface_init (XdpBackgroundIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (Background, background, XDP_TYPE_BACKGROUND_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_TYPE_BACKGROUND, background_iface_init));
+
+
+typedef enum { UNSET, NO, YES, ASK } Permission;
+
+static Permission
+get_permission (const char *app_id)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) out_perms = NULL;
+  g_autoptr(GVariant) out_data = NULL;
+  const char **permissions;
+
+  if (!xdp_impl_permission_store_call_lookup_sync (get_permission_store (),
+                                                   PERMISSION_TABLE,
+                                                   PERMISSION_ID,
```

Comment:
Why are you using this index here? It seems like we're basically returning the entire table every time. Is that going to scale? Can't you use the app id as index instead?

Reply:
I guess you need to get the permissions for all the apps in the checking case, to avoid a lot of dbus calls. Still, it seems non-ideal.

Reply:
> In some earlier permutation of this we had some leeway for being in the background to avoid the "race condition" where an app was just starting in parallel with the app background check. Like, allow the app to be in the background for one cycle, and only ask the second time we check. I think we need something like that to be robust.

Still the case. I added a comment to make that clear.

Reply:
I'll leave this to a future cleanup to have better permission plumbing, across all portals